### PR TITLE
Blog post: Import Path FAQ

### PR DESCRIPTION
**Pull request by cameel**
This is still work in progress, but I'm submitting it already to get some early feedback. I have the later sections sketched out but I still need to turn my outline into nice text and add examples.

It's pretty long, though given how much I had to cut from the path docs, this is to be expected. I wonder if should maybe split it into smaller posts though?

### CODE REVIEW

File: _posts/2021-08-16-import-path-faq.md [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+And the details can get complicated if your program has to work on a variety of platforms with completely different filesystem conventions or even not having a filesystem at all (e.g. a browser).
+
+This is the very reason why the Solidity compiler takes a hands-off approach to the problem.
+The [Remix IDE](https://remix.ethereum.org/) is widely known for supporting Github URLs but what many developers do not know is that you can use literally any string to identify your file.
+Internally the compiler makes very few assumptions about that string which, ironically, sometimes leads to misunderstandings when it clashes with programmer's expectations.
+In the end the source code in most cases comes from a filesystem and users do not refer to it using abstract identifiers.
+They expect the compiler to map their local filesystem content to its own internal structure in a consistent way regardless of the platform.
+Even when using [Standard JSON](https://docs.soliditylang.org/en/latest/using-the-compiler.html#compiler-input-and-output-json-description) format, allows tools to access that structure directly, the burden just shifts to the tool.
```
Comment (chriseth) [inline _posts/2021-08-16-import-path-faq.md:19]: grammar is off

**Comment by cameel** [comment]
I have just pushed the remaining text. The post is now complete but I want to make another pass through it before I mark it as ready for review. I know there are some typos and grammar errors and it needs some polishing. Especially in the later sections where I focused on finally getting it all out with the intention of fixing small details later.

The feedback that would be most useful at this point is whether the overall structure makes sense. Looking at section titles and just lightly skimming through text and examples - does it look ok? I decided to use the FAQ format in the hope of making it more interesting and addressing real problems I've seen rather than just throwing a long monotonous list of examples at the reader, but the downside of this approach is that I ended up re-explaining some stuff that's already in the docs.

**Comment by franzihei** [comment]
My two gwei to the current state:
- Overall structure seems fine
- The post is __really__ long. Not many people will read such a long post. If it is not possible to explain the stuff shorter that's ok I guess, but it usually helps to offer different levels of details aka, have a digest of what is important to know somewhere in the beginning (like a tl;dr) and then the more detailed stuff
- The title doesn't necessarily motivate to read more / is not really intriguing.. maybe we can tweak it a bit?

**Comment by cameel** [comment]
Yeah, I'm fully aware that it's way too long for its own good, just I'm not sure how it can serve its purpose without that level of detail. Part of the reason path handling is notoriously buggy and inconsistent, is that no one really wants to dig that deep into all these details and special cases. There's tons of them and it's very tedious. Laying them out like this makes it easier than having to figure them out on your own. Though it definitely won't beat the ease of not having to worry about them :)

I'm going to try to make it more concise, which should help a bit but that probably won't be more than 10-20% without cutting stuff. The problem really is that the post is all detail :) It's not meant to give a high level overview because that's what the docs I wrote earlier already do. Maybe I should just assume that the reader has fully read the docs? That would let me omit some explanations but it would also make it hard to read the post on its own.

Maybe the best solution would be to split it into two parts? I wanted to avoid it because I don't think the topic is interesting enough to people that they'd want to read multiple articles about it, but two parts is not that much and maybe preferable to having one monster article anyway :) I think I could fairly cleanly split it into one post focusing on compiler's VFS and another dealing only with the import callback and the actual filesystem.

> The title doesn't necessarily motivate to read more / is not really intriguing.. maybe we can tweak it a bit?

True, the title is a bit plain. Here's some more ideas:
- "Dealing with import problems in Solidity"
- "Common problems with import paths"
- "Deep dive into compiler's virtual filesystem"
- "Imports and the filesystem"
- "The import callback and the VFS"
- "More than you ever wanted to know about import paths in Solidity"
- "The twisted paths of Solidity" (a really bad pun :P)

Also maybe you have some ideas yourself?

File: _posts/2021-08-16-import-path-faq.md [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+---
+
+Paths are one of the things that "just work" as long as you're not too particular about the details.
+If you can just hand off the path to a library, it's no longer your problem.
+And the details can get complicated if your program has to work on a variety of platforms with completely different filesystem conventions or even not having a filesystem at all (e.g. a browser).
+
+This is the very reason why the Solidity compiler takes a hands-off approach to the problem.
+The [Remix IDE](https://remix.ethereum.org/) is widely known for supporting Github URLs but what many developers do not know is that you can use literally any string to identify your file.
+Internally the compiler makes very few assumptions about that string which, ironically, sometimes leads to misunderstandings when it clashes with programmer's expectations.
+In the end the source code in most cases comes from a filesystem and users do not refer to it using abstract identifiers.
```
Comment (Marenz) [inline _posts/2021-08-16-import-path-faq.md:17]:
```suggestion
In the end the source code comes in most cases from a filesystem and users don't refer to it using abstract identifiers.
```

File: _posts/2021-08-16-import-path-faq.md [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+They expect the compiler to map their local filesystem content to its own internal structure in a consistent way regardless of the platform.
+Even when using [Standard JSON](https://docs.soliditylang.org/en/latest/using-the-compiler.html#compiler-input-and-output-json-description) format, allows tools to access that structure directly, the burden just shifts to the tool.
```
Comment (Marenz) [inline _posts/2021-08-16-import-path-faq.md:19]:
```suggestion
And while our [Standard JSON](https://docs.soliditylang.org/en/latest/using-the-compiler.html#compiler-input-and-output-json-description) format allows tools to access that structure directly, the burden just shifts to the tool.
```

File: _posts/2021-08-16-import-path-faq.md [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+
+The details of the whole mechanism have recently been documented in [Import Path Resolution](https://docs.soliditylang.org/en/latest/path-resolution.html) but many of the consequences of the documented behavior may not be apparent.
```
Comment (Marenz) [inline _posts/2021-08-16-import-path-faq.md:21]:
```suggestion
The details of the whole mechanism have recently been added to our [Import Path Resolution documentation](https://docs.soliditylang.org/en/latest/path-resolution.html) but many of the consequences of the documented behavior may not be apparent.
```

File: _posts/2021-08-16-import-path-faq.md [hunk 5]
```diff
@@ -0,0 +1,526 @@
...
+This blog post is meant to be a more complete exploration of problems you might encounter when dealing with paths in Solidity.
+It will show you specific corner cases, explain why they happen and tell what you can do to avoid them.
+
+### Q: What's wrong with relative paths in Solidity? Aren't `import "file.sol"` and `import "./file.sol"` the same thing?
```
Comment (Marenz) [inline _posts/2021-08-16-import-path-faq.md:25]:
```suggestion
### Q: What's wrong with relative paths in Solidity? Isn't `import "file.sol"` the same as `import "./file.sol"`?
```

Yours is grammatically correct as well. Dunno I felt mine reads nicer, but just a suggestion.

File: _posts/2021-08-16-import-path-faq.md [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+No, in Solidity these two statements are processed very differently.
+`file.sol` is actually considered absolute in compiler's virtual filesystem.
```
Comment (Marenz) [inline _posts/2021-08-16-import-path-faq.md:27]:
```suggestion
`file.sol` is actually considered absolute in the compiler's virtual filesystem.
```

**Review by Marenz — CHANGES_REQUESTED** [review]

**Pull request closed without merging**